qe/drivers.py

```diff
@@ -1707,3 +1866,682 @@ def selection_step(
+        Used in building Hamiltonian to be diagonalized."""
+
+        return self.Hamiltonian_one_electron_driver.H(self.psi_i, self.psi_j)
+
+    @cached_property
+    def H_full(self):
+        if self.comm is None:
+            return self.H
+        else:
+            """Build full one-electron Hamiltonian H = [H_1; ...; H_N]."""
+            H_i = self.H  # Each process builds local Hamiltonian
+            recvcounts = None
+            recvcounts = np.array(
+                self.comm.gather(self.local_size * self.full_size, root=0)
+            )  # No. of elements received by each process
+
+            H_full = np.zeros((self.full_size, self.full_size), dtype="float")
+            # Master process gathers and sends full two-electron Hamiltonian
+            self.comm.Gatherv(H_i, (H_full, recvcounts), root=0)
+            self.comm.Bcast(np.array(H_full, dtype="float"), root=0)  # TODO: changed from bcast
+
+            return H_full
+
+    def D_i(self):
+        """
+        Build local diagonal of one-electron Hamiltonian in determinant-driven fashion.
+        Needed for pre-conditioning step of Davidson's method.
+        :return D_i: n_local dimensional numpy array
+        """
+        D_i = np.zeros(self.local_size, dtype="float")  # Save diagonal as a numpy vector
+        for j, det in enumerate(
+            self.psi_i
+        ):  # Iterate through determinants in local portion of wavefunction
+            D_i[j] = self.Hamiltonian_one_electron_driver.H_ii(det)
+        return D_i
+
+    def implicit_1e_matrix_product(self, V_k):
+        """Function to implicitly compute matrix-matrix product W_ik = H_i * V_k by building the local Hamiltonian
+        `on the fly', where H_i is the one-electron portion of the local Hamiltonian.
+
+        :param H_i: local (self.local_size \times n) one-electron portion of the Hamiltonian, never explicitly stored
+        :param V_k: full column basis for trial subspace (n \times k), as a numpy array
+
+        :return W_ik: locally computed chunk of matrix-matrix product (self.local_size \times k), as a numpy array
+        """
+        try:
+            k = V_k.shape[1]
+        except IndexError:  # Handle case when V is a vector
+            k = 1
+
+        W_ik = np.zeros(
+            (self.local_size, k), dtype="float"
+        )  # Pre-allocate space for local brick of matrix-matrix product
+        # One-electron part
+        for I, det_I in enumerate(self.psi_i):
+            for J, det_J in enumerate(self.psi_j):
+                try:
+                    W_ik[I, :] += (
+                        self.Hamiltonian_one_electron_driver.H_ij(det_I, det_J) * V_k[J, :]
+                    )  # Update row I of W_ik
+                except IndexError:  # If V_k is a single vector.
+                    W_ik[I] += (
+                        self.Hamiltonian_one_electron_driver.H_ij(det_I, det_J) * V_k[J]
+                    )  # Update row I of W_ik
```

Comment:
This is ugly (you have 2 code path for the vector, matrix case), the trick if you want to support matrix and vector as an input argument is to transform the vector into a matrix as the beginning of the function